### DISCUSSION TITLE
Save changes to related rows only when parent row is saved

### DISCUSSION THREAD

**Discussion by fcapurso** [Budibase/budibase#8446] [Help]
Hi!

I've been using Budibase for about a week now, trying to build an alternative to an app I have already developed using another tool, and so far I like it a lot!

There's one specific problem I haven't found a good solution for though. In the app I'm building I will need to create "recipes", which will contain a list of ingredients and respective amounts. Because ingredients can be used in multiple recipes and they will be linked to products for which I need to maintain an inventory, these will be held in a separate table from the recipes table and the link between the two is realised by means of an intermediate table (which is also where the amounts are stored).
On a UI level, I'd like there to be a page that allows to create (or modify) a recipe from where I can also manage the ingredient list.

So far I was able to create a page that allows me to do this, but whatever change a make to an ingredient is saved immediately to the DB, which makes the user experience not that great in case the user should want to abort the recipe creation/modification half way and revert all changes made while the recipe page was open.

I can imagine this is a rather recurring used case (it can also occur for things like orders / order items or invoices / invoice lines).
Basically I'd like to keep all the changes made in some "temporary" storage until the user either saves the recipe or cancels / leaves the page.

Can somebody suggest a way to approach this?

Thanks a lot in advance,
Filippo

**Comment by melohagan** [comment]
Hey Filippo,

You can save to multiple tables in one button click as shown here:

- https://docs.budibase.com/docs/form-steps
- https://docs.budibase.com/docs/chaining-actions#example-creating-and-linking-two-table-entries-in-one-form

**Reply by fcapurso** [reply]
Hi @melohagan,

Thanks a lot for your answer!
That provides me a way to save two rows in different tables at once indeed, which can be very useful for other use cases I have (the second link is particularly useful).
However for the use case above I'd need to collect data from the user about a number of related rows (not possible to know in advance how many) before I save them all. This would have to be applicable both to create and update actions.
Any further suggestion on how this could be achieved?

**Reply by fcapurso** [reply]
To make it more concrete: here you can see how such a screen would look like.
<img width="300" alt="image" src="https://user-images.githubusercontent.com/3785896/200302905-88495637-6044-40fb-a78d-86842b24bf9c.png">
The way I'm doing it for the moment is to popup another form for every ingredient the user wants to edit, but the changes have to be saved on the spot (when the user leaves the popup and goes back to the recipe form). Which is not very user friendly because if you changed your mind about modifying your recipe, it's not enough to leave the recipe form without pressing on Save, and there's also no way to "undo" your changes.

**Reply by melohagan** [reply]
If the number of relationships is dynamic/unknown, then you'll need action looping which currently is only available in the automations: https://docs.budibase.com/docs/looping

**Reply by fcapurso** [reply]
That could work indeed. But I think the number of steps needs to be know in advance? Or could I collect data on an arbitrary number of related rows (ingredients) in the same screen/step?
